src/util/ALE.sol

```diff
@@ -0,0 +1,235 @@
+        // Permit data
+        Permit calldata _permit,
+        // Optional helper data in case collateral needs to be transformed
+        bytes calldata _helperData
+    ) external payable nonReentrant {
+        if (markets[_sellTokenAddress].market == address(0))
+            revert MarketNotSetForCollateral(_sellTokenAddress);
+
+        IMarket market = IMarket(markets[_sellTokenAddress].market);
+        IERC20 collateral = IERC20(markets[_sellTokenAddress].collateral);
+        
+        dola.mint(address(this), _value);
+        dola.approve(address(market), _value);
+
+        market.repay(msg.sender, _value);
+
+        // withdraw amount from ZERO EX quote
+        market.withdrawOnBehalf(
+            msg.sender,
+            _collateralAmount,
+            _permit.deadline,
+            _permit.v,
+            _permit.r,
+            _permit.s
+        );
+
+        // If there's an helper contract, the collateral has to be transformed
+        if (markets[_sellTokenAddress].helper != address(0)) {
+            //Approve collateral to be transformed to sellToken
+            collateral.approve(
+                markets[_sellTokenAddress].helper,
+                _collateralAmount
+            );
+            // Collateral amount is now transformed into sellToken
+            _collateralAmount = ITransformHelper(
+                markets[_sellTokenAddress].helper
+            ).transformFromCollateral(_collateralAmount, _helperData);
+            // Approve sellToken for spender
+            IERC20(_sellTokenAddress).approve(_spender, _collateralAmount);
+        } else {
+            // If there's no helper the withdrawn collateral is the sellToken
+            if (address(collateral) == address(0)) revert CollateralNotSet();
+            // Approve collateral to be swapped
+            collateral.approve(_spender, 0); // in some cases like USDT we need to first set it to zero and the approve (TODO review this)
+            collateral.approve(_spender, _collateralAmount);
+        }
+
+        uint256 dolaBalBefore = dola.balanceOf(address(this));
+
+        // Call the encoded swap function call on the contract at `swapTarget`,
+        // passing along any ETH attached to this function call to cover protocol fees.
+        // NOTE: This will swap the collateral or helperCollateral for DOLA
+        (bool success, ) = exchangeProxy.call{value: msg.value}(_swapCallData);
+        if (!success) revert SwapFailed();
+
+        uint256 dolaBalAfter = dola.balanceOf(address(this));
+       
+        if (dolaBalAfter - dolaBalBefore < _value) revert DOLAInvalidRepay();
+
+        dola.burn(_value);
+
+        // Send any DOLA leftover to the sender after burning (in case the collateral withdrawn and swapped exceeds the value to burn)
+        dola.transfer(msg.sender, dola.balanceOf(address(this))); 
+
+        // Refund any unspent protocol fees to the sender.
+        payable(msg.sender).transfer(address(this).balance);
```

Comment:
Not sure if an optimization or not, depends on how likely it is there will be a leftover balance.
```suggestion
        if(address(this).balance > 0) payable(msg.sender).transfer(address(this).balance);
```

Reply:
Fixed in [21265a2](https://github.com/InverseFinance/FiRM/pull/52/commits/21265a282ef6b19fa53c46e5b48fe83f48880686)